### PR TITLE
Improved duck array wrapping

### PR DESCRIPTION
**Pull request by slevang**
- [x] Issue: #7848
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

Companion to #9776.

My attempts to use xarray-wrapped jax arrays turned up a bunch of limitations of our duck array wrapping. Jax is probably worst case of the major duck array types out there, because it doesn't implement either `__array_function__` or `__array_ufunc__` to intercept numpy calls. Cupy, sparse, and probably others do, so calling `np.func(cp.ndarray)` generally works fine, but this usually converts your data to numpy with jax.

 A lot of this was just grepping around for hard-coded `np.` cases that we can easily dispatch to `duck_array_ops` versions. I image some of these changes could be controversial, because a number of them (notably `nanmean/std/var`, `pad`, `quantile`, `einsum`, `cross`) aren't technically part of the standard. See https://github.com/pydata/xarray/issues/8834#issuecomment-1998340481 for discussion about the nan-skipping aggregations.

It feels like a much better user experience though to try our best to dispatch to the correct backend, and error if the function isn't implemented, rather than blindly calling numpy. And practically, all major array backends that are feasible to wrap today (`cupy`, `sparse`, `jax`, `cubed`, `arkouda`, ...?) implement all of these functions.

To test, I just ran down the API list and ran most functions to see if we maintain proper wrapping. Prior to the changes here, I had 28 jax failures and 9 cupy failures, while all (non-xfailed) ones now pass.

Basically everything works except the interp/missing methods which have a lot of specialized code. Also a few odds and ends like polyfit and rank.

### CODE REVIEW

File: xarray/core/duck_array_ops.py [hunk 1]
```diff
@@ -133,15 +122,20 @@ def fail_on_dask_array_input(values, msg=None, func_name=None):
...
-sliding_window_view = _dask_or_eager_func(
-    "sliding_window_view",
-    eager_module=np.lib.stride_tricks,
-    dask_module=dask_array_compat,
-    dask_only_kwargs=("automatic_rechunk",),
-    numpy_only_kwargs=("subok", "writeable"),
-)
+
+def sliding_window_view(array, window_shape, axis=None, **kwargs):
+    # TODO: some libraries (e.g. jax) don't have this, implement an alternative?
```
Comment (slevang) [inline xarray/core/duck_array_ops.py:127]: This is one of the biggest outstanding bummers of wrapping jax arrays. There is apparently [openness](https://github.com/jax-ml/jax/issues/11354) to adding this as an API (even though it would not offer any performance benefit in XLA). But given this is way outside the API standard, whether it makes sense to implement a general version within xarray that doesn't rely on stride tricks.

File: xarray/tests/test_duck_array_wrapping.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+import xarray as xr
+
+# TODO: how to test these in CI?
```
Comment (slevang) [inline xarray/tests/test_duck_array_wrapping.py:7]: note: I just noticed [xarray-array-testing](https://github.com/xarray-contrib/xarray-array-testing)

**Comment by lucascolley** [comment]
see also https://github.com/data-apis/array-api/issues/621 for the higher level discussion of `nan` reductions

File: xarray/core/array_api_compat.py [hunk 3]
```diff
@@ -42,3 +44,29 @@ def result_type(*arrays_and_dtypes, xp) -> np.dtype:
...
+    def _get_single_namespace(x):
+        if hasattr(x, "__array_namespace__"):
+            return x.__array_namespace__()
+        elif isinstance(x, array_type("cupy")):
+            # special case cupy for now
+            import cupy as cp
+
+            return cp
+        else:
+            return np
```
Comment (lucascolley) [inline xarray/core/array_api_compat.py:59]: you probably want to wrap array-api-compat's `array_namespace` here, see https://github.com/scipy/scipy/blob/ec30b43e143ac0cb0e30129d4da0dbaa29e74c34/scipy/_lib/_array_api.py#L118-L152 for what we do in SciPy.

Comment (slevang) [reply xarray/core/array_api_compat.py:59]: Happy to go that route, I did actually try that but array-api-compat doesn't handle a bunch of things we end up passing through this ([scalars](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py#L562), index wrappers, etc) so it would require some careful prefiltering. 

The only things this package effectively wraps that don't have `__array_namespace__` are cupy, dask, and torch. I tried the torch wrapper but it doesn't pass our [as_compatible_data](https://github.com/pydata/xarray/blob/main/xarray/core/variable.py#L265) check because the wrapper object itself doesn't have `__array_namespace__` or `__array_function__` :confused:

Comment (slevang) [reply xarray/core/array_api_compat.py:59]: Looks like [this](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py#L234) would handle the array-like check well, but would require adding this as a core xarray dependency to use it in `as_compatible_data`. Not sure if there is any appetite for that.

Comment (slevang) [reply xarray/core/array_api_compat.py:59]: I played around with this some more.

Getting an object that is compliant from the perspective of `array_api_compat.is_array_api_obj` to work through the hierarchy of xarray objects requires basically swapping in a bunch of more restrictive `hasattr(__array_namespace__)` checks for this function. Not too bad.

I was able to get things to the point that xarray can wrap `torch.Tensor`, which is pretty cool. But the reality is that xarray relies on a lot of functionality beyond the Array API, so this just doesn't work in any practical sense. Similarly, swapping in the more restrictive `cupy.array_api` for the main `cupy` namespace causes all kinds of things to break.

It seems [torch](https://github.com/pytorch/pytorch/issues/58743) has had little movement on supporting the standard since 2021. From xarray's perspective, `cupy` implements everything we need, except for the actual `__array_namespace__` attribute declaring compatibility, which seems to be [planned for v14](https://github.com/cupy/cupy/issues/8269).

So while this compat module is nice in theory, I don't think it's very useful for xarray. cupy, jax, and sparse are good to go without it, and we only need a single special case for cupy to fetch its `__array_namespace__`.

**Comment by slevang** [comment]
I looked back at `cupy-xarray`. Now that this duck array stuff all works pretty well, I'm wondering how people feel about adding official `DataArray/Dataset` methods analogous to [this](https://github.com/xarray-contrib/cupy-xarray/blob/main/cupy_xarray/accessors.py#L70), but in a generalized way. Something like:

```python
def as_array_type(self, asarray: callable, **kwargs) -> Self:
# e.g. ds.to_array_type(jnp.asarray, device="gpu")

def is_array_type(self, array: type) -> bool:
# e.g. ds.is_array_type(cp.ndarray)
```

File: xarray/core/duck_array_ops.py (hunk 1 above)
Comment (slevang) [reply xarray/core/duck_array_ops.py:127]: I have something that works pretty well with [this](https://github.com/jax-ml/jax/issues/3171#issuecomment-631826501) style of gather operation. But only in a `jit` context where XLA can work its magic. So I guess this is better left to the specific library to implement, or the user.

**Comment by dcherian** [comment]
> methods analogous to [this](https://github.com/xarray-contrib/cupy-xarray/blob/main/cupy_xarray/accessors.py?rgh-link-date=2024-11-20T15%3A47%3A08Z#L70), but in a generalized way.

:+1: I though I'd seen an array API method for converting between compliant array types, but I can't find it now

**Comment by slevang** [comment]
Compliant namespaces should now implement `from_dlpack` which is generally the recommended conversion protocol. So I suppose we could instead pass the namespace and hard code it to use that:

```python
def to_namespace(self, xp: ModuleType, **kwargs) -> Self:
    xp.from_dlpack(self.data)
       
# e.g. ds.to_namespace(cp)
```

But this actually doesn't work for cupy, since they're quite stringent about implicit device transfers:
```
TypeError: CPU arrays cannot be directly imported to CuPy. Use `cupy.array(numpy.from_dlpack(input))` instead.
```

Also sparse [doesn't have this at all](https://github.com/pydata/sparse/blob/c0b2b4506de77434c0e3cd8b7ed3e04b407dca54/ci/Numba-array-api-xfails.txt#L35), and it wouldn't be clear whether you want a `COO`, `DOK`, etc.

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -8678,16 +8678,23 @@ def _integrate_one(self, coord, datetime_unit=None, cumulative=False):
                     coord_names.add(k)
             else:
                 if k in self.data_vars and dim in v.dims:
+                    # cast coord data to duck array if needed
+                    if isinstance(v.data, array_type("cupy")):
+                        coord_data = duck_array_ops.get_array_namespace(v.data).asarray(
+                            coord_var.data
+                        )
+                    else:
+                        coord_data = coord_var.data
```
Comment (shoyer) [inline xarray/core/dataset.py:8687]: It looks like this could be encapsulated into an `as_like_array()` helper function?

I'd like to keep the explicit array namespace stuff out of most xarray functions.

File: xarray/core/variable.py [hunk 5]
```diff
@@ -860,12 +860,17 @@ def _getitem_with_mask(self, key, fill_value=dtypes.NA):
             # we need to invert the mask in order to pass data first. This helps
             # pint to choose the correct unit
             # TODO: revert after https://github.com/hgrecco/pint/issues/1019 is fixed
-            data = duck_array_ops.where(np.logical_not(mask), data, fill_value)
+            # cast mask to any duck array type
+            if type(mask) is not type(data):
+                mask = duck_array_ops.get_array_namespace(data).asarray(mask)
+            data = duck_array_ops.where(
+                duck_array_ops.logical_not(mask), data, fill_value
+            )
```
Comment (shoyer) [inline xarray/core/variable.py:868]: this could also use `as_like_array()`

**Comment by slevang** [comment]
I think this is in pretty good shape now, except the question of whether to attempt any of [this integration testing](https://github.com/pydata/xarray/pull/9798/files#diff-6b85bb2bc68c311c9b042341bec6647ec2e953ac740378682deba9c62d68597f) in CI. That could also be punted to [xarray-array-testing](https://github.com/xarray-contrib/xarray-array-testing) (@keewis @TomNicholas)

**Comment by shoyer** [comment]
The new `is_array_type` and `as_array_types` methods seem worthy of broader discussion. I would leave them out of this PR for now, so they don't hold up mergin it.

**Comment by slevang** [comment]
Sure I'll split that off. Agree the naming and API could use some discussion.

**Comment by dcherian** [comment]
> Basically everything works except the interp/missing methods which have a lot of specialized code. 

interp will need scipy to support array-api objects which I think they're working on.

> Also a few odds and ends like polyfit and rank.

yeah funny that the linalg extension does not include `lstsq`.

---

More generally, this is great work @slevang. thanks!

File: xarray/tests/test_duck_array_wrapping.py (hunk 2 above)
Comment (dcherian) [reply xarray/tests/test_duck_array_wrapping.py:7]: we do install pint in CI if that helps.

**Comment by lucascolley** [comment]
> yeah funny that the linalg extension does not include lstsq.

[From 2021](https://github.com/data-apis/array-api/pull/240#issue-972003513):

> Initial discussion on linalg.lstsq API design can be found in https://github.com/data-apis/array-api/issues/227. There, it was held that the current API design posed undue burden on specification implementers due to API complexity (returning solutions, ranks, residuals, and singular values) and violated various linear algebra API design principles (particularly orthogonality).
> 
> Based on ensuing discussion in consortium meetings, consensus was found in simply removing the linalg.lstsq API altogether. This API finds low usage among downstream libraries (see [data](https://github.com/data-apis/array-api-comparison/blob/bb9922d42d2cc9d040bb0b60612f345356442fc2/data/common_apis_ranks.csv#L191)) and has several variants due to specialized use cases (see SciPy).
> 
> Should this or similar functionality be found to be desirable in the future, specification can happen in a future revision of the standard.

You aren't the first to hit it, reading that issue. Perhaps you could raise it again :)

**Review by dcherian — APPROVED** [review]

**Comment by slevang** [comment]
fwiw, `jax` and `cupy` both have `linalg.lstsq`, although it sounds like `solve` would be the right way to go.

I just didn't tackle this function because it is defined on Dataset, and has probably 15 calls to different numpy functions, followed by a loop over variables. So it seemed tricky to handle in case we have data vars with mixed namespaces or other edge cases.

File: xarray/core/duck_array_ops.py [hunk 6]
```diff
@@ -172,7 +166,8 @@ def isnull(data):
         )
     ):
         # these types cannot represent missing values
-        return full_like(data, dtype=bool, fill_value=False)
+        dtype = xp.bool_ if hasattr(xp, "bool_") else xp.bool
```
Comment (shoyer) [inline xarray/core/duck_array_ops.py:169]: I believe `xp.bool_` is only needed for compatibility with NumPy 1.x releases. Can we make that clearer here, e.g., by using `np.bool_` and adding a comment about NumPy versions?

File: xarray/tests/test_duck_array_wrapping.py (hunk 2 above)
Comment (shoyer) [reply xarray/tests/test_duck_array_wrapping.py:7]: Cupy requries a GPU, but we could add `jax` into Xarray's CI suite in GitHub actions, at least.

**Review by shoyer — APPROVED** [review]

File: xarray/tests/test_duck_array_wrapping.py (hunk 2 above)
Comment (keewis) [reply xarray/tests/test_duck_array_wrapping.py:7]: the only thing we'll want to avoid is testing compat with too many array types in `xarray`'s CI. I guess I'll need to make progress on `xarray-array-testing` to avoid that, though.

File: xarray/core/duck_array_ops.py (hunk 6 above)
Comment (slevang) [reply xarray/core/duck_array_ops.py:169]: Will do, but also `cupy` only has `bool_` at the moment. But If I passed built-in `bool` to this I think I got a cupy error. It was a weird range from like numpy 1.2x until 2 that they deprecated `bool` and only had `bool_`, then brought it back in 2.0 :shrug:

File: xarray/tests/test_duck_array_wrapping.py (hunk 2 above)
Comment (slevang) [reply xarray/tests/test_duck_array_wrapping.py:7]: `jax` would be pretty easy to add. The CPU jaxlib package is only like 90MB. And it's a nice test case since it very closely matches the numpy API, but only implements `__array_namespace__`, and not `__array_function__` or `__array_ufunc__`.

Comment (shoyer) [reply xarray/tests/test_duck_array_wrapping.py:7]: I agree, we should add JAX CPU. See https://github.com/pydata/xarray/tree/main/ci/requirements for the config files. Feel free to do this in this PR or a follow-up.

File: xarray/tests/test_duck_array_wrapping.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+        },
+    },
+    "cupy": {
+        "attrs": {
+            "array": "ndarray",
+            "constructor": "asarray",
+        },
+        "xfails": {"quantile": "no nanquantile"},
+    },
+    "pint": {
```
Comment (slevang) [inline xarray/tests/test_duck_array_wrapping.py:26]: Added pint and sparse, and more consolidated xfail lists. Some of these are probably desired behavior, like all the methods that return bool or int not returning a `pint.Quantity`, but it was easiest to just xfail them.

File: ci/requirements/environment.yml [hunk 8]
```diff
@@ -49,3 +49,5 @@ dependencies:
   - toolz
   - typing_extensions
   - zarr
+  - pip:
+      - jax # no way to get cpu-only jaxlib from conda if gpu is present
```
Comment (slevang) [inline ci/requirements/environment.yml:53]: Trying to install via conda on my local machine with a GPU automatically gave all the CUDA libs, which seems like unnecessary bloat for dev environments. If anyone knows a way around this, happy to move it back to conda. I also got a failure on windows, I think for trying to pip install anything, so left it off of there.

**Comment by dcherian** [comment]
This seems like a great step forward. Thanks @slevang
